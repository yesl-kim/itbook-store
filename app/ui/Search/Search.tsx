'use client'

import { useSearchParams } from 'next/navigation'
import { useRouter } from 'next/navigation'
import { useCallback } from 'react'
import { SubmitHandler, useForm } from 'react-hook-form'

interface SearchForm {
  keyword: string
}

const PATH = '/books'
export default function Search() {
  const searchParams = useSearchParams()
  const { push } = useRouter()
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<SearchForm>({
    defaultValues: {
      keyword: searchParams.get('query')?.toString(),
    },
  })

  const onSubmit: SubmitHandler<SearchForm> = ({ keyword }) => {
    const params = new URLSearchParams()
    params.set('query', keyword)
    params.set('page', '1')
    push(`${PATH}?${params.toString()}`)
    const keywords = parseKeyword(keyword)
    // TODO: keywords를 여러개 전달해야함 ex) ?q=a&q=b...
  }

  return (
    <div>
      <form
        className="relative flex flex-1 flex-shrink-0"
        onSubmit={handleSubmit(onSubmit)}
      >
        <input
          {...register('keyword', {
            required: true,
            validate: validateMaxKeyword,
          })}
          role="search"
          placeholder="검색어를 입력해주세요"
          className="peer block w-full rounded-md border border-gray-200 py-[9px] pl-10 text-sm outline-2 placeholder:text-gray-500"
        />
        <input type="submit" value="검색" />
      </form>
      <p>{errors.keyword?.message}</p>
    </div>
  )
}

const parseKeyword = (keyword: string): string[] => {
  const seperatorRegex = /[\|\-]/
  return keyword.split(seperatorRegex)
}

const validateMaxKeyword = (input: string) => {
  const keywords = parseKeyword(input)
  if (keywords.length > 2) {
    return '키워드는 최대 2개까지 입력할 수 있습니다'
  }
  return true
}
