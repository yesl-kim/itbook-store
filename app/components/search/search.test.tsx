import { fireEvent, render, screen } from '@testing-library/react'
import useEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'
import mockRouter from 'next-router-mock'

jest.mock('next/navigation', () => jest.requireActual('next-router-mock'))

import Search from './search'

const PATH = '/books'
describe('search component', () => {
  it('검색하면 입력값이 쿼리 파라미터("query")에 담겨 "books/"로 이동한다', () => {
    mockRouter.push('/')
    render(<Search />)

    const input = 'keyword'
    const search = screen.getByRole('search')
    fireEvent.change(search, { target: { value: input } })

    expect(mockRouter).toMatchObject({
      asPath: `${PATH}?query=${input}`,
      pathname: PATH,
      query: { query: input },
    })
  })
  it('입력값이 주어지지 않은 경우 검색은 무시된다', () => {})
  it('"|", "-"로 구분되는 검색 키워드는 최대 2개까지만 입력될 수 있다. 넘어갈 경우 검색이 무시된다', () => {})
  test('snapshot', () => {})
})
