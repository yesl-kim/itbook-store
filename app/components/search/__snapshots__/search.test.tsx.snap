// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search component snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="relative flex gap-2 w-full max-w-[900px]"
      >
        <input
          class="peer block w-full rounded-md border border-gray-200 py-[9px] pl-10 text-sm outline-2 placeholder:text-gray-50"
          name="query"
          placeholder="검색어를 입력해주세요"
          role="search"
        />
        <input
          type="submit"
          value="검색"
        />
      </form>
      <p
        class="text-sm text-red-400"
      />
    </div>
  </body>,
  "container": <div>
    <form
      class="relative flex gap-2 w-full max-w-[900px]"
    >
      <input
        class="peer block w-full rounded-md border border-gray-200 py-[9px] pl-10 text-sm outline-2 placeholder:text-gray-50"
        name="query"
        placeholder="검색어를 입력해주세요"
        role="search"
      />
      <input
        type="submit"
        value="검색"
      />
    </form>
    <p
      class="text-sm text-red-400"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
